File: examples/cccbenchplushwloc.cpp
#include <iostream>
#include <filesystem>

#include "sys-sage.hpp"
#include "parsers/cccbench.hpp"

void usage(char* argv0)
{
    std::cerr << "usage: " << argv0 << " <hwloc xml path> <cccbench csv path>" << std::endl;
    return;
}

int main(int argc, char *argv[])
{
    string xmlPath;
    std::string cccPath;
    if(argc == 3){
        xmlPath = argv[1];
        cccPath = argv[2];
    }
    else{
        usage(argv[0]);
        return 1;
    }

    //create root Topology and one node
    Topology* topo = new Topology();
    Node* n = new Node(topo, 1);

    if(parseHwlocOutput(n, xmlPath) != 0) { //adds topo to a next node
        usage(argv[0]);
        return 1;
    }
    cout << "-- End parseHwlocOutput" << endl;

    cout << "Total num HW threads: " << topo->CountAllSubcomponentsByType(SYS_SAGE_COMPONENT_THREAD) << endl;

    cout << "---------------- Printing the whole tree ----------------" << endl;
    topo->PrintSubtree(2);
    cout << "----------------                     ----------------" << endl;
    
    

    parseCccbenchOutput(n, cccPath);


    auto allcores = new vector<Component *>();
    topo->GetAllSubcomponentsByType(allcores, SYS_SAGE_COMPONENT_CORE);    
    //auto allcores = topo->GetAllChildrenByType(SYS_SAGE_COMPONENT_CORE);

    for(auto c0 : *allcores)
        for(auto c1 : *allcores)
            cout << endl << c0->GetId() << " " << c1->GetId() << endl;

    cout << "---------------- Printing all DataPaths ----------------" << endl;
    n->PrintAllDataPathsInSubtree();
    cout << "----------------                        ----------------" << endl;
    delete topo;
    delete n;
    return 0;
}

File: examples/cccbenchplushwloc.py
import py_sys_sage as pysage 
import sys
import os
def usage(argv0):
    print(f"usage: {argv0} <hwloc xml path> <cccbench csv path>", file=sys.stderr)

def main():
    if len(sys.argv) != 3:
        usage(sys.argv[0])
        return 1

    xmlPath = sys.argv[1]
    cccPath = sys.argv[2]

    # create root Topology and one node
    topo = pysage.Topology()
    n = pysage.Node(topo, 1)

    if pysage.parseHwlocOutput(n, xmlPath) != 0:  # adds topo to a next node
        usage(sys.argv[0])
        return 1
    print("-- End parseHwlocOutput")

    print("Total num HW threads:", topo.CountAllSubcomponentsByType(pysage.COMPONENT_THREAD))

    print("---------------- Printing the whole tree ----------------")
    topo.PrintSubtree(2)
    print("----------------                     ----------------")

    pysage.parseCccbenchOutput(n, cccPath)

    allcores = []
    topo.GetAllSubcomponentsByType(allcores, pysage.COMPONENT_CORE)

    for c0 in allcores:
        for c1 in allcores:
            print(f"\n{c0.GetId()} {c1.GetId()}")

    print("---------------- Printing all DataPaths ----------------")
    n.PrintAllDataPathsInSubtree()
    print("----------------                        ----------------")

    return 0

if __name__ == "__main__":
    main()

